### DISCUSSION TITLE
Why was postgres:// dropped / deprecated?

### DISCUSSION THREAD

**Discussion by dchevell** [sqlalchemy/sqlalchemy#6423] [Ideas]
Hi team,

I see in SQLAlchemy 1.4 we dropped the URI short form `postgres://` after a long deprecation, now requiring the long form `postgresql://`.

I can't find anything that points to the original _reason_ for this. Everything just points back to the work that removed it, but I can't find any supporting explanation. As far as I can tell, Postgres itself continues to support and document both forms with no indication that the short form is deprecated by them.

The reason for asking is that I've raised this in my org as something that we may need to fix in certain layers of our stack, but when asked "why are SQLAlchemy deprecating something that seems supported?" I haven't got a reasonable answer. If you could shed any light or context it would be helpful.

Cheers

Dave

**Comment by CaselIT** [comment]
Hi,

Please see https://github.com/sqlalchemy/sqlalchemy/issues/6083#issuecomment-801478013

As for the why, I think (I'm just supposing here, I was not around in 2010 when the original deprecation was added) the reason is that sqlalchemy associates a database url prefix with a package name of the dialect. So it does not support, without special logic, two prefixes pointing to the same package. This special logic was known to be used only in the `postgres://` case for a feature that was deprecated from 11 years. So it was removed.

**Reply by CaselIT** [reply]
I think that the fact that so many people did not notice a warning that was there since 2009, really illustrates how broken are the default python warning filter more than anything else.

**Reply by dchevell** [reply]
I found that, but it also doesn't really address the question. I know a number of issues have been opened asking how to fix it, but I'm only really interested in understanding why.

I'm receiving feedback from internal stakeholders that they consider SQLAlchemy's decision "wrong", in not supporting something that _is_ supported by other libraries, frameworks, and by Postgres itself. I'm just looking to have the other side of the argument to have a more valuable discussion. 

The special logic theory is an interesting one, but dropping it seems like an odd choice if it's dropping something that is _supposed_ to work, as documented by Postgres. It would mean SQLAlchemy's implementation is now incomplete. I'm sure there's a counterargument to this but I don't know what it is.

**Reply by dchevell** [reply]
Also worth noting that even though this has been warned about for a while, a lot of places are working with tooling that spans many languages and frameworks, and making this change across an organisation for the sake of one library requires some additional evidence and explanation. I'll take anything I can get. 

I'm a bit disappointed about closing this as it suggests nobody else is going to look at it to answer the question - and it seems that whilst this topic has come up plenty of times in the context of break/fix or mistaken bug reports, nothing has addressed the underlying reasoning. I think that's worth digging into for such a significant library in the python ecosystem.

**Reply by CaselIT** [reply]
> I'm receiving feedback from internal stakeholders that they consider SQLAlchemy's decision "wrong", in not supporting something that is supported by other libraries, frameworks, and by Postgres itself. I'm just looking to have the other side of the argument to have a more valuable discussion.
> The special logic theory is an interesting one, but dropping it seems like an odd choice if it's dropping something that is supposed to work, as documented by Postgres. It would mean SQLAlchemy's implementation is now incomplete. I'm sure there's a counterargument to this but I don't know what it is.

I don't think SQLAlchemy has ever claimed to accept default dbapi strings, since it also supports `+drivername` to use a particular dbapi, deciding for a default implementation for each db in case that's missing. As for the db prefix, the dbapi is also the name of a python module, that gets loaded by parsing the string. I'm not sure there is any expectation that this form is a standard used by other libraries. (I may be wrong here).
The deprecation is just the removal of a particular resolution logic in sqlalchemy dialect loading code, not that necessary that url format was deprecated (I may have wrongly posted it, but re-reading pg documentation it seems the two are valid aliases)

> Also worth noting that even though this has been warned about for a while, a lot of places are working with tooling that spans many languages and frameworks, and making this change across an organisation for the sake of one library requires some additional evidence and explanation. I'll take anything I can get.

There are various possible solutions to avoid having to make the change organization wide, like replacing the string connection before creating the engine or using a custom creator function. 

In any case the sensible think would have been to open an issue regarding the deprecation in the 11 years it was raising a deprecation warning. Since any no such issue was ever reported, the developers of SQLAlchemy thought, I guess wrongly, that it was not a problem for anyone. We have no way of knowing that people are relying on deprecated features if we do not get any feedback about it.

> I'm a bit disappointed about closing this as it suggests nobody else is going to look at it to answer the question - and it seems that whilst this topic has come up plenty of times in the context of break/fix or mistaken bug reports, nothing has addressed the underlying reasoning. I think that's worth digging into for such a significant library in the python ecosystem.

It's not closed, I moved it to a discussion, where we are moving issues that have nothing actionable (yet) on it.

**Comment by CaselIT** [comment]
For the sake of completess, @zzzeek replied to this here https://github.com/sqlalchemy/sqlalchemy/issues/6083#issuecomment-831604283

> 
> 
> > > can we "dupe" all these ? there were other issues opened w/ people surprised by the name removal right?
> > 
> > 
> > @zzzeek is there anywhere that captures the _why_ for deprecation?
> 
> the reason the name was changed, which goes back to [here](https://docs.sqlalchemy.org/en/14/changelog/changelog_06.html#change-0ce9ee480a5a66cb9f77e0207f6dfc70), is because I had called the database dialect "postgres" because my original work with this database was in the 1995-ish kind of time, when that's what it was called. Then around SQLAlchemy 0.5 or so, people told me hey, this database is actually called "PostgreSQL" now. So I realized I had made a mistake, and i changed the name.
> 
> That's "why the change" but at that point the name "postgres" was legacy and we added a deprecation warning at that point. "Why the deprecation" is because, it is normal to deprecate the old name of something when you fix the name to be something new.
> 
> > database URI's are something broader than SQLAlchemy and python,
> 
> SQLAlchemy URLs are not broader than SQLAlchemy. They are specific to SQLAlchemy, do not look like anyone else's URLs (and they are more URLs than just URIs, I think) and they only work with SQLAlchemy.
> 
> > and I've run into situations in my org where we have orchestration layers injecting DB URI's into applications across many languages and frameworks - all of which continue to support the short form.
> 
> I can't imagine using a single database URL without modification across many different application platforms. I guess because PostgreSQL's libpq actually has a URL format that is fairly well known , there is some feasibility to this but certainly in the larger database world this is not a thing at all. Like here is pymysql.connect: https://pymysql.readthedocs.io/en/latest/modules/connections.html there is no URL there. URLs for databases are not a universally available thing.
> 
> > Postgres' own docs still list it as valid.
> 
> I'm pretty sure this is up in the discussion here, that's the PostgreSQL database. SQLAlchemy URLs are not libpq URIs.
> 
> > I'm not here to argue the point (it's done now, and was deprecated long ago), but when asking internal teams to change to the long form at a layer that affects all languages/stacks/frameworks, the argument that one ORM decided to drop it doesn't hold water without some context as to why.
> 
> I'm not following why that matters.
> 
> > I'm currently getting pushback from teams that consider SQLAlchemy "wrong" for dropping support for something that is still documented as supported by Postgres itself.
> 
> I don't understand the nature of this "pushback". You didnt make this change, I did, and it's already done, so...what are they pushing against? if these teams are in the business of using libpq URIs as a universal database connectivity format then shouldn't there be "pushback" that they aren't supporting MySQL, Oracle, SQL Server as well ? If these teams are saying something like, "well we shouldnt support SQLAlchemy because they made this fickle decision"...that's too bad but honestly it makes them look a lot **more** fickle to make such a big deal over a change like this.
> 
> > I don't have a sufficient counter-argument, other than "switching to the long form would make my own life easier" stuck_out_tongue
> 
> what are....the two sides to this argument? "sqlalchemy is wrong" / "sqlalchemy is right" ...uhhh, water is wet / water is moist? why does it matter?

And here https://github.com/sqlalchemy/sqlalchemy/issues/6083#issuecomment-831654538

> 
> 
> OK, I figured I'd do some research on this libpq URI issue. Because first I noticed one thing, then I had kind of a weird suspicion, and then...the suspicion turned out to be ... true!
> 
> The feature was only added relatively recently in 9.2: https://www.postgresql.org/docs/release/9.2.0/ which is two years after we deprecated "postgres". it was also six years after SQLAlchemy's first release so our URL format predates theirs significantly.
> 
> Yet, as we see in the actual commit at [postgres/postgres@b035cb9#diff-66bf98128702c8fd0063e6989416d0d1739556d12f772d38a1845b83bd0acada](https://github.com/postgres/postgres/commit/b035cb9db7aa7c0f28581b23feb10d3c559701f6#diff-66bf98128702c8fd0063e6989416d0d1739556d12f772d38a1845b83bd0acada) , they seemed to focus on "postgres" but also added the comment (because there just wasnt enough irony already) "Oh, it also supports postgresql:// but that's probably just an accident." . The URI format looks so much like ours that I crawled their dev list from that time to find where this was first proposed and sure enough our format was part of their consideration:
> 
> https://www.postgresql.org/message-id/4F45253C.4030409%402ndQuadrant.com
> 
> > First thing:  the URI prefix.  It is possible to connect using a URI in
> > Python+SQL Alchemy, which was mentioned before as not too relevant due
> > to their also requiring a driver name.  As documented at
> > http://docs.sqlalchemy.org/en/latest/core/engines.html and demonstrated
> > at http://packages.python.org/Flask-SQLAlchemy/config.html , it is
> > possible to leave off the driver part of the connection string.  That
> > assumes the default driver, such that postgresql:// does the same as
> > postgresql+psycopg2:// , sensibly.  That means we absolutely have an
> > installed base of URI speaking developers split between postgresql://
> > (Python) and postgres:// (Ruby).  Given that, there really isn't a
> > useful path forward that helps out all those developers without
> > supporting both prefixes.  That's where this left off before, I just
> > wanted to emphasize how clear that need seems now.
> 
> So what we have above is exactly the rationale used by PG's developers to support both "postgresql://" as well as "postgres://". **to be compatible with SQLAlchemy**. So there you have it. Their URL format **did not exist** when we deprecated "postgres://" and their use of "postgresql://" was at least somewhat based on their observations of us. fin
